src/Components/Components/src/SupplyParameterFromPersistentComponentStateValueProvider.cs

```diff
@@ -0,0 +1,281 @@
+                        Debug.Assert(written == 0);
+                        GrowBuffer(ref pool, ref keyBuffer);
+                    }
+                    else
+                    {
+                        currentBuffer = currentBuffer[..written];
+                    }
+                }
+            }
+            else
+            {
+                var keySpan = ResolveKeySpan(key);
+                var wroteKey = false;
+                while (!wroteKey)
+                {
+                    currentBuffer = keyBuffer[preKey.Length..];
+                    wroteKey = Encoding.UTF8.TryGetBytes(keySpan, currentBuffer, out var written);
+                    if (!wroteKey)
+                    {
+                        // It is really unlikely that we will enter here, but we need to handle this case
+                        Debug.Assert(written == 0);
+                        // Since this is utf-8, grab a buffer the size of the key * 4 + the preKey size
+                        // this guarantees we have enough space to encode the key
+                        GrowBuffer(ref pool, ref keyBuffer, keySpan.Length * 4 + preKey.Length);
+                    }
+                    else
+                    {
+                        currentBuffer = currentBuffer[..written];
+                    }
+                }
+            }
+
+            keyBuffer = keyBuffer[..(preKey.Length + currentBuffer.Length)];
+
+            var hashSucceeded = SHA256.TryHashData(keyBuffer, keyHash, out _);
+            Debug.Assert(hashSucceeded);
+            return Convert.ToBase64String(keyHash);
+        }
+        finally
+        {
+            if (pool != null)
+            {
+                ArrayPool<byte>.Shared.Return(pool, clearArray: true);
+            }
+        }
+    }
+
+    private static ReadOnlySpan<char> ResolveKeySpan(object? key)
+    {
+        if (key is IFormattable formattable)
+        {
+            var keyString = formattable.ToString("", CultureInfo.InvariantCulture);
+            return keyString.AsSpan();
+        }
+        else if (key is IConvertible convertible)
+        {
+            var keyString = convertible.ToString(CultureInfo.InvariantCulture);
+            return keyString.AsSpan();
+        }
+        return default;
+    }
+
+    private static void GrowBuffer(ref byte[]? pool, ref Span<byte> keyBuffer, int? size = null)
+    {
+        var newPool = pool == null ? ArrayPool<byte>.Shared.Rent(size ?? 2048) : ArrayPool<byte>.Shared.Rent(pool.Length * 2);
+        keyBuffer.CopyTo(newPool);
+        keyBuffer = newPool;
+        if (pool != null)
+        {
+            ArrayPool<byte>.Shared.Return(pool, clearArray: true);
+        }
+        pool = newPool;
+    }
+
+    private static object? GetSerializableKey(ComponentState componentState)
+    {
+        if (componentState.LogicalParentComponentState is not { } parentComponentState)
```

Comment:
I would have expected this to use `ParentComponentState` instead of `LogicalParentComponentState`, since the physical parent is what would supply the key not the logical parent. Unless I'm misremembering how this works. Can you clarify if that was considered? If for example there's some test case showing it needs to be the `LogicalParentComponentState` then I would be wrong but I haven't got to the test cases yet :)

Reply:
I thought logical is what we wanted in case we wanted it to work with `<Section>` too, but I'll revisit this

Reply:
> I thought logical is what we wanted in case we wanted it to work with <Section> too, but I'll revisit this

In the `<Section>` case, what key do you want to build? One that regards the component as being a descendant of the outlet, or of the source?

Reply:
I've switched this to use `ParentComponent` state